core/src/impl/Kokkos_HostSharedPtr.hpp

```diff
@@ -0,0 +1,170 @@
+
+#include <Kokkos_Macros.hpp>
+#include <Kokkos_Atomic.hpp>
+
+#include <functional>
+
+namespace Kokkos {
+namespace Impl {
+
+template <typename T>
+class MaybeReferenceCountedPtr {
+ protected:
+  MaybeReferenceCountedPtr(T* value_ptr)
+      : m_value_ptr(value_ptr), m_control(nullptr) {}
+
+  template <class Deleter>
+  MaybeReferenceCountedPtr(T* value_ptr, Deleter deleter)
+      : m_value_ptr(value_ptr), m_control(new Control{std::move(deleter), 1}) {}
+
+ public:
+  KOKKOS_FUNCTION MaybeReferenceCountedPtr(
+      MaybeReferenceCountedPtr&& other) noexcept
+      : m_value_ptr(other.m_value_ptr), m_control(other.m_control) {
+    other.m_value_ptr = nullptr;
+    other.m_control   = nullptr;
+  }
+
+  KOKKOS_FUNCTION MaybeReferenceCountedPtr(
+      const MaybeReferenceCountedPtr& other) noexcept
+      : m_value_ptr(other.m_value_ptr), m_control(other.m_control) {
+#ifdef KOKKOS_ACTIVE_EXECUTION_MEMORY_SPACE_HOST
+    if (m_control) Kokkos::atomic_add(&(m_control->m_counter), 1);
+#endif
+  }
+
+  KOKKOS_FUNCTION MaybeReferenceCountedPtr& operator=(
+      MaybeReferenceCountedPtr&& other) noexcept {
+    if (&other != this) {
+      cleanup();
+      m_value_ptr       = other.m_value_ptr;
+      other.m_value_ptr = nullptr;
+      m_control         = other.m_control;
+      other.m_control   = nullptr;
+    }
+    return *this;
+  }
+
+  KOKKOS_FUNCTION MaybeReferenceCountedPtr& operator=(
+      const MaybeReferenceCountedPtr& other) noexcept {
+    if (&other != this) {
+      cleanup();
+      m_value_ptr = other.m_value_ptr;
+      m_control   = other.m_control;
+#ifdef KOKKOS_ACTIVE_EXECUTION_MEMORY_SPACE_HOST
+      if (m_control) Kokkos::atomic_add(&(m_control->m_counter), 1);
+#endif
+    }
+    return *this;
+  }
+
+  KOKKOS_FUNCTION ~MaybeReferenceCountedPtr() { cleanup(); }
+
+  KOKKOS_FUNCTION T* get() const noexcept { return m_value_ptr; }
+  KOKKOS_FUNCTION T& operator*() const noexcept { return *get(); }
+  KOKKOS_FUNCTION T* operator->() const noexcept { return get(); }
+
+  // checks if the stored pointer is not null
+  KOKKOS_FUNCTION explicit operator bool() const noexcept {
+    return get() != nullptr;
+  }
+
+  // checks whether the MaybeReferenceCountedPtr manages the lifetime of the
+  // object
+  KOKKOS_FUNCTION bool owner() const noexcept { return m_control != nullptr; }
+
+  KOKKOS_FUNCTION int use_count() const noexcept {
+    return m_control ? m_control->m_counter : 0;
+  }
+
+ private:
+  KOKKOS_FUNCTION void cleanup() noexcept {
+#ifdef KOKKOS_ACTIVE_EXECUTION_MEMORY_SPACE_HOST
+    // If m_counter is set, then this instance is responsible for managing the
+    // objects pointed to by m_counter and m_value_ptr.
+    if (m_control == nullptr) return;
+    int const count = Kokkos::atomic_fetch_sub(&(m_control->m_counter), 1);
+    if (count == 1) {
+      (m_control->m_deleter)(m_value_ptr);
+      m_value_ptr = nullptr;
+      delete m_control;
+      m_control = nullptr;
+    }
```

Comment:
I slightly prefer `if(owner())` and then the rest of the body in a block. Also the `const` on a stack variable seems unnecessary to me, but I don't care that much I guess. Also, `KOKKOS_EXPECTS(m_control->m_deleter)`?